python/mxnet/gluon/contrib/data/audio/datasets.py

```diff
@@ -0,0 +1,171 @@
+    transform : callable, default None
+        A function that takes data and label and transforms them
+    has_csv: default True
+        If True, it means that a csv file has filename and its corresponding label
+        If False, we have folder like structure
+    train_csv: str, default None
+        If has_csv is True, train_csv should be populated by the training csv filename
+    file_format: str, default '.wav'
+        The format of the audio files(.wav, .mp3)
+    skip_rows: int, default 0
+        While reading from csv file, how many rows to skip at the start of the file to avoid reading in header
+
+    Attributes
+    ----------
+    synsets : list
+        List of class names. `synsets[i]` is the name for the integer label `i`
+    items : list of tuples
+        List of all audio in (filename, label) pairs.
+    """
+    def __init__(self, root, has_csv=False, train_csv=None, file_format='.wav', skip_rows=0):
+        self._root = os.path.expanduser(root)
+        self._exts = ['.wav']
+        self._format = file_format
+        self._has_csv = has_csv
+        self._train_csv = train_csv
+        self._list_audio_files(self._root, skip_rows=skip_rows)
+
+
+    def _list_audio_files(self, root, skip_rows=0):
+        """
+            Populates synsets - a map of index to label for the data items.
+            Populates the data in the dataset, making tuples of (data, label)
+        """
+        self.synsets = []
+        self.items = []
+        if not self._has_csv:
+            for folder in sorted(os.listdir(root)):
+                path = os.path.join(root, folder)
+                if not os.path.isdir(path):
+                    warnings.warn('Ignoring %s, which is not a directory.'%path, stacklevel=3)
+                    continue
+                label = len(self.synsets)
+                self.synsets.append(folder)
+                for filename in sorted(os.listdir(path)):
+                    file_name = os.path.join(path, filename)
+                    ext = os.path.splitext(file_name)[1]
+                    if ext.lower() not in self._exts:
+                        warnings.warn('Ignoring %s of type %s. Only support %s'%(filename, ext, ', '.join(self._exts)))
+                        continue
+                    self.items.append((file_name, label))
+        else:
+            data_tmp = []
+            label_tmp = []
+            skipped_rows = 0
+            with open(self._train_csv, "r") as traincsv:
+                for line in traincsv:
+                    skipped_rows = skipped_rows + 1
+                    if skipped_rows <= skip_rows:
+                        continue
+                    filename = os.path.join(root, line.split(",")[0])
+                    label = line.split(",")[1].strip()
+                    if label not in self.synsets:
+                        self.synsets.append(label)
+                    data_tmp.append(os.path.join(self._root, line.split(",")[0]))
```

Comment:
should we first see if the folder has synset file? I mean, user can specify synset file. And, logic here will prepare synset when user did not provide one. WDYT?

Reply:
The current logic will overwrite any synset.txt file currently in the folder. I suppose this is okay as when the user creates an AudioFolderDataset, it can be a completely new folder or data itself. Here the assumption is that the audio/wav files are present in the folder like structure (siren/0.wav...) or all files are present in a folder with a csv file containing filename to label mapping. The process of generating the synset should be internal to the logic. I may be wrong but exposing this to user is something I am not sure of. I referred to the following issue for this:
https://github.com/apache/incubator-mxnet/issues/5163